### PR TITLE
Add flag to reverse errors

### LINKED ISSUE #3860: Add a flag to forc to reverse the order of printed errors
Requested by @nfurfaro in a recent sway sync. Since most of the time, lots of errors propagates from the first error (meaning that once the first error is fixed, all of them goes away) it would make sense to be able to revert the order so that users do not need to scroll.

### PR DESCRIPTION
**Pull request by gr00vytvniks**
closes #3860

### CODE REVIEW

**Comment by gr00vytvniks** [comment]
hey @kayagokalp I made the changes you requested, I tried running `forc build` after adding some errors to `examples/counter` and I got this error.
```sh
error
  --> /Users/mitchmackert/Code/sway/sway-lib-core/src/ops.sw:11:9
   |
 9 | 
10 |     fn add(self, other: Self) -> Self {
11 |         __add(self, other)
   |         ^^^^^ Identifiers cannot begin with a double underscore, as that naming convention is reserved for compiler intrinsics.
12 |     }
13 | }
   |
____
```
would you mind giving me steps to test this correctly? I think I am looking at this the wrong way

**Comment by kayagokalp** [comment]
hey @gr00vytvniks thanks for the PR, the error looks odd. Can you make rebase/merge master you shouldn't have any lock file is the counter folder as we recently changed that folder to be a workspace, [see](https://github.com/FuelLabs/sway/tree/master/examples/counter). This type of core compilation errors can be seen with master forc depending on a released version of std/core. Can you make sure that in `examples/counter/Forc.toml` you actually depend on local std. You should have something like this in the `Forc.toml`:

```toml
[dependencies]
std = { path = "../../sway-lib-std" }TOML
``` 

You should be able to test this with a template project by adding a couple of compile errors. Let me know if you are unsure/stuck 👍

**Comment by gr00vytvniks** [comment]
> hey @gr00vytvniks thanks for the PR, the error looks odd. Can you make a rebase/merge master as you shouldn't have any lock file is the counter folder as we recently changed that folder to be a workspace, [see](https://github.com/FuelLabs/sway/tree/master/examples/counter). This type of core compilation errors can be seen with master forc depending on a released version of std/core. Can you make sure that in `examples/counter/Forc.toml` you actually depend on local std. You should have something like this in the `Forc.toml`:
> 
> ```toml
> [dependencies]
> std = { path = "../../sway-lib-std" }
> ```
> 
> You should be able to test this with a template project by adding a couple of compile errors. Let me know if you are unsure/stuck 👍

Hey @kayagokalp, The error came from a collision between an old fuelup binary and a cargo binary.

**Comment by gr00vytvniks** [comment]
Hey @kayagokalp,  I added a debug build profile like the book instructed but it is giving me this warning.

```
 ~/Code/sway/examples/counter/ [gr00vytvniks/reverse-errors*] forc build
  WARNING! unused manifest key: build-profiles
  WARNING! unused manifest key: build-profiles
  WARNING! unused manifest key: build-profiles
  WARNING! unused manifest key: build-profiles
  WARNING! unused manifest key: build-profiles
  WARNING! unused manifest key: build-profiles
  Compiled library "core".
  Compiled library "std".
[forc-util/src/lib.rs:243] reverse_errors = false
warning
  --> /Users/mitchmackert/Code/sway/examples/counter/src/main.sw:19:9
   |
17 | 
18 |         storage.counter = value;
19 |         value
   |         ----- This cast, from integer type of width sixty four to integer type of width thirty two, will lose precision.
20 |     }
   |
____

warning
  --> /Users/mitchmackert/Code/sway/examples/counter/src/main.sw:26:9
   |
24 | 
25 |         storage.counter = incremented;
26 |         incremented
   |         ----------- This cast, from integer type of width sixty four to integer type of width eight, will lose precision.
27 |     }
28 | }
   |
____

warning
  --> /Users/mitchmackert/Code/sway/examples/counter/src/main.sw:12:5
   |
10 | 
11 | storage {
12 |     counter: u64 = 0,
   |     ------- This storage declaration is never accessed and can be removed.
13 | }
   |
____

error
  --> /Users/mitchmackert/Code/sway/examples/counter/src/main.sw:17:42
   |
15 | 
16 |     #[storage(write)]
17 |     fn initialize_counter(value: u64) -> u32 {
   |                                          ^^^ expected: u64 
found:    u32 
help:     The definition of this function must match the one in the ABI "TestContract" declaration.
18 |         storage.counter = value;
19 |         value
   |
____

error
  --> /Users/mitchmackert/Code/sway/examples/counter/src/main.sw:23:42
   |
21 | 
22 |     #[storage(read, write)]
23 |     fn increment_counter(amount: u64) -> u8 {
   |                                          ^^ expected: u64 
found:    u8 
help:     The definition of this function must match the one in the ABI "TestContract" declaration.
24 |         let incremented = storage.counter + amount;
25 |         storage.counter = incremented;
   |
____

  Aborting due to 2 errors.
Error: Failed to compile counter

```

**Comment by gr00vytvniks** [comment]
I went ahead and just changed the default in `debug()` to `true` and this was the output :) 
```
error
  --> /Users/mitchmackert/Code/sway/examples/counter/src/main.sw:23:42
   |
21 | 
22 |     #[storage(read, write)]
23 |     fn increment_counter(amount: u64) -> u8 {
   |                                          ^^ expected: u64 
found:    u8 
help:     The definition of this function must match the one in the ABI "TestContract" declaration.
24 |         let incremented = storage.counter + amount;
25 |         storage.counter = incremented;
   |
____

error
  --> /Users/mitchmackert/Code/sway/examples/counter/src/main.sw:17:42
   |
15 | 
16 |     #[storage(write)]
17 |     fn initialize_counter(value: u64) -> u32 {
   |                                          ^^^ expected: u64 
found:    u32 
help:     The definition of this function must match the one in the ABI "TestContract" declaration.
18 |         storage.counter = value;
19 |         value
   |
____

  Aborting due to 2 errors.
Error: Failed to compile counter
```

File: examples/counter/Forc.lock
Comment (kayagokalp) [inline examples/counter/Forc.lock:14]: We should not have individual lock files as examples folder is a workspace now. Did you manually add this file? How did it produced?

**Review by kayagokalp — CHANGES_REQUESTED** [review]
We should not revert it in any build profile imo. This is not something everyone would be using it is more niche command that some people may end-up using. 

So we should not revert by default as it would be unexpected for the users. We should be able to use this feature with the flag `forc build --reverse-errors` etc.

File: examples/counter/Forc.lock
Comment (gr00vytvniks) [reply examples/counter/Forc.lock:14]: I figured that by adding the `reverse_errors` field to the `BuildProfile` the idiomatic way for the user would be able to add it to their build profile instead of passing a flag, that way they can just run `forc build`. 
The lock file is being generated probably because I need to update.

Comment (gr00vytvniks) [reply examples/counter/Forc.lock:14]: @kayagokalp

Comment (kayagokalp) [reply examples/counter/Forc.lock:14]: Hey @gr00vytvniks, I am waiting you to remove the lock file to do a final review as it shouldn't be there. After removing it and building the example again, if it is generated again we might have a problem and we should deal with it before merging this one.

Comment (gr00vytvniks) [reply examples/counter/Forc.lock]: Hey @kayagokalp I removed the lock file and it looks like I just had an outdated forc version. 👍

**Review by nfurfaro — DISMISSED** [review]
Thanks!

File: forc-pkg/src/pkg.rs
Comment (eureka-cpu) [inline forc-pkg/src/pkg.rs:1726]: `terse_mode` is defined twice here

File: forc-pkg/src/pkg.rs
Comment (eureka-cpu) [inline forc-pkg/src/pkg.rs:2299]: Looks like you're returning `Err` twice here

File: forc-util/src/lib.rs
Comment (eureka-cpu) [inline forc-util/src/lib.rs:240]: remove this `dbg!` macro if this is ready

**Comment by eureka-cpu** [comment]
Looks like there's a problem with the formatting of the `examples/counter/Forc.toml` file. Also some of the `should_fail` e2e tests are failing incorrectly (not sure if it's related, may just need to re-run the job).

File: forc-pkg/src/manifest.rs
Comment (kayagokalp) [inline forc-pkg/src/manifest.rs:612]:
```suggestion
            reverse_errors: false,
```
I think we should not reverse the errors by default even if we are using a debug profile. This is pretty specialized feature and users might be surprised if errors are reverted out of the box. Users can pass the flag if they want to have it explicitly.

Also this is why some/all of the should fail tests failing as we are looking for specific errors with them reversing the order breaks the tests.

**Review by kayagokalp — CHANGES_REQUESTED** [review]
Hey @gr00vytvniks, looking good! Left a simple thing to address which may fix your CI issues as well (I suppose e2e-tests failing due to this change)

File: forc-pkg/src/manifest.rs
Comment (eureka-cpu) [reply forc-pkg/src/manifest.rs:612]: Does `forc build --reverse-errors` work, or does this only work when creating a custom build profile?

Comment (eureka-cpu) [reply forc-pkg/src/manifest.rs]: Willing to approve once this is resolved. I think it would be okay to keep this as a feature that you have to specify in a build profile, since it's likely that those who would use it would always want to use it. I think it would be redundant to have both, as it would defeat the purpose of having the option for custom build profiles. If there is a flag it may lead to people using it under a command alias, instead.

Comment (eureka-cpu) [reply forc-pkg/src/manifest.rs]: cc @nfurfaro thoughts?

Comment (nfurfaro) [reply forc-pkg/src/manifest.rs]: I've never used a build profile yet... a flag would be sufficient for me, but don't see any harm in supporting both.

**Review by kayagokalp — COMMENTED** [review]
Hey @gr00vytvniks current changes looks perfect! I only have a single nit and a very simple addition request so that your new feature is useable through our CLI and lets merge this afterwards. 

To be able to do that we are just missing cli bindings for the new `reverse_error` bool, we could have a field under `Print` in `forc/src/cli/shared.rs` and it could look like this:

```rust
/// Options related to printing stages of compiler output.
#[derive(Args, Debug, Default)]
pub struct Print {
    /// Print the generated Sway AST (Abstract Syntax Tree).
    #[clap(long)]
    pub ast: bool,
    /// Print the computed Sway DCA graph.
    #[clap(long)]
    pub dca_graph: bool,
    /// Print the finalized ASM.
    ///
    /// This is the state of the ASM with registers allocated and optimisations applied.
    #[clap(long)]
    pub finalized_asm: bool,
    /// Print the generated ASM.
    ///
    /// This is the state of the ASM prior to performing register allocation and other ASM
    /// optimisations.
    #[clap(long)]
    pub intermediate_asm: bool,
    /// Print the generated Sway IR (Intermediate Representation).
    #[clap(long)]
    pub ir: bool,
    /// Output the time elapsed over each part of the compilation process.
    #[clap(long)]
    pub time_phases: bool,
    /// Output build errors and warnings in reverse order.
    #[clap(long)]
    pub reverse_order: bool,
}
```

and in `forc/src/ops/forc_build.rs` we need to add the cli flag which would make that place something like:

```rust
fn opts_from_cmd(cmd: BuildCommand) -> pkg::BuildOpts {
    pkg::BuildOpts {
        pkg: pkg::PkgOpts {
            path: cmd.build.pkg.path,
            offline: cmd.build.pkg.offline,
            terse: cmd.build.pkg.terse,
            locked: cmd.build.pkg.locked,
            output_directory: cmd.build.pkg.output_directory,
            json_abi_with_callpaths: cmd.build.pkg.json_abi_with_callpaths,
        },
        print: pkg::PrintOpts {
            ast: cmd.build.print.ast,
            dca_graph: cmd.build.print.dca_graph,
            finalized_asm: cmd.build.print.finalized_asm,
            intermediate_asm: cmd.build.print.intermediate_asm,
            ir: cmd.build.print.ir,
            reverse_order: cmd.build.print.ast
        },
        time_phases: cmd.build.print.time_phases,
        minify: pkg::MinifyOpts {
            json_abi: cmd.build.minify.json_abi,
            json_storage_slots: cmd.build.minify.json_storage_slots,
        },
        build_profile: cmd.build.profile.build_profile,
        release: cmd.build.profile.release,
        error_on_warnings: cmd.build.profile.error_on_warnings,
        binary_outfile: cmd.build.output.bin_file,
        debug_outfile: cmd.build.output.debug_file,
        build_target: cmd.build.build_target,
        tests: cmd.tests,
        member_filter: Default::default(),
    }
}
```

File: forc-util/src/lib.rs
Comment (gr00vytvniks) [reply forc-util/src/lib.rs:240]: Thanks for the review @kayagokalp! I'll get right on this :)

File: forc-plugins/forc-client/src/op/deploy.rs
Comment (kayagokalp) [inline forc-plugins/forc-client/src/op/deploy.rs:166]: Hmm this looks like an error should we be using cmd.build.print.reverse_order?

File: forc-plugins/forc-client/src/op/run.rs
Comment (kayagokalp) [inline forc-plugins/forc-client/src/op/run.rs:162]: ^

File: forc/src/cli/commands/test.rs
Comment (kayagokalp) [inline forc/src/cli/commands/test.rs:171]: ^

File: forc/src/ops/forc_build.rs
Comment (kayagokalp) [inline forc/src/ops/forc_build.rs:27]: ^

**Review by kayagokalp — CHANGES_REQUESTED** [review]
Thanks for the updates! I think we have some copy paste errors leaked here but otherwise lgtm 👍

File: forc/src/ops/forc_build.rs
Comment (gr00vytvniks) [reply forc/src/ops/forc_build.rs:27]: Gotcha! Must have slipped past me but I will fix right away :)

**Comment by eureka-cpu** [comment]
Happy to approve this once merge conflicts are resolved and @kayagokalp gives his approval

**Comment by eureka-cpu** [comment]
As a follow up PR, you could update the documentation so that developers will know how to use this feature. ie as part of a `BuildProfile` or just via `--reverse-order`(Not sure what name you guys landed on).

File: forc-pkg/src/manifest.rs
Comment (eureka-cpu) [inline forc-pkg/src/manifest.rs:228]: I think this field `experimental_private_modules` was removed at some point as it isn't on `master` anymore. I also think the `serde` annotation above `reverse_results` was intended for that field and can be removed as well. cc @kayagokalp

File: forc-pkg/src/manifest.rs
Comment (eureka-cpu) [inline forc-pkg/src/manifest.rs:672]: It can be removed here as well

File: forc-pkg/src/manifest.rs
Comment (eureka-cpu) [inline forc-pkg/src/manifest.rs:690]: and here

**Review by eureka-cpu — CHANGES_REQUESTED** [review]
Haven't tested yet but there's one thing that seems out of place if you wouldn't mind addressing it. After that I'll pull this branch and test it out!

File: forc-pkg/src/manifest.rs
Comment (gr00vytvniks) [reply forc-pkg/src/manifest.rs:228]: Ah I see. Fixing :)

File: forc-pkg/src/manifest.rs
Comment (eureka-cpu) [inline forc-pkg/src/manifest.rs:226]: I don't think you need to have the `#[serde(default)]` annotation here. Here's the doc that explains what it does, though I think it was just leftover from the field you had removed here.

Comment (gr00vytvniks) [reply forc-pkg/src/manifest.rs]: Just got that done and committed, thank you for the patience :)

**Review by eureka-cpu — DISMISSED** [review]
🚢

**Comment by JoshuaBatty** [comment]
Thanks for this. just some last merge conflicts @gr00vytvniks

**Review by JoshuaBatty — APPROVED** [review]

**Review by eureka-cpu — APPROVED** [review]

**Review by kayagokalp — APPROVED** [review]

**Pull request merged**
